# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/builtin/FinalizationRegistryObject.h
# Commit: 5a6684922b98
# Full Hash: 5a6684922b980d32b05bd7210cfadc34e734a845
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-03-21 21:28:09
# Regressor Bug: 1623973
# File Overlap Count: 2
# Description:
#   Bug 1623973 - Make FinalizationRegistryObject active record set and registrations map values weak r=sfink
#   
#   This removes tracing for the values of the registration map and the active record set and sweeps them instead. This requires maintaining a map of FinalizationRegistryObjects on the zone.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D67677
# ==============================================================================

diff -r c4d2ca8f78b7 -r 5a6684922b98 js/src/builtin/FinalizationRegistryObject.h
--- a/js/src/builtin/FinalizationRegistryObject.h	Thu Mar 19 16:42:42 2020 +0000
+++ b/js/src/builtin/FinalizationRegistryObject.h	Fri Mar 20 21:57:22 2020 +0000
@@ -135,8 +135,10 @@
 using FinalizationRecordVector =
     GCVector<HeapPtr<FinalizationRecordObject*>, 1, js::ZoneAllocPolicy>;
 
-// A JS object that wraps a FinalizationRecordVector. Used as the values in the
-// registration weakmap.
+// A JS object containing a vector of FinalizationRecordObjects, which holds the
+// records corresponding to the registrations for a particular registration
+// token. These are used as the values in the registration weakmap. The contents
+// of the vector are weak references and are not traced.
 class FinalizationRecordVectorObject : public NativeObject {
   enum { RecordsSlot = 0, SlotCount };
 
@@ -153,6 +155,8 @@
   bool append(HandleFinalizationRecordObject record);
   void remove(HandleFinalizationRecordObject record);
 
+  void sweep();
+
  private:
   static const JSClassOps classOps_;
 
@@ -192,6 +196,8 @@
   void setQueuedForCleanup(bool value);
   void setCleanupJobActive(bool value);
 
+  void sweep();
+
   static bool cleanupQueuedRecords(JSContext* cx,
                                    HandleFinalizationRegistryObject registry,
                                    HandleObject callback = nullptr);